### PR TITLE
feat: allow saving CFP submissions as drafts without required fields

### LINKED ISSUE #2317: Allow saving draft CFP submissions without completing required fields
In the Call for Proposals (CFP) flow, speakers are currently required to complete all **required fields** before they can save their submissions as a draft.

This makes it difficult for speakers to
• Start a proposal and return later
• Incrementally work on longer submissions
• Save partial ideas or notes

In practice, required field validation should apply to **final submission**, not to saving a draft.

## Expected behaviour

• Speakers can **save a proposal as draft** even if required fields are not completed
• Validation of required fields is enforced **only when submitting** the proposal
• Draft proposals can be edited and completed later

## Current behaviour

• Required fields block both
– saving a draft
– submitting a proposal
• Users cannot save incomplete proposals

## Proposed solution

• Separate validation logic for
– **Save draft**
– **Submit proposal**
• Skip required field validation when saving drafts
• Enforce full validation only on submission

## Scope

• CFP submission flow
• Validation logic for proposal forms
• Draft state handling
• Frontend and backend validation consistency

## Acceptance criteria

• Draft proposals can be saved with missing required fields
• Submitting a proposal without required fields is still blocked
• Clear distinction between draft and submitted state
• No regressions in existing CFP submission behaviour

## Notes

This improves usability for speakers and aligns the CFP flow with common submission patterns used in conference management systems.

### PR DESCRIPTION
**Pull request by sarafarajnasardi**
Fixes #2317
before

https://github.com/user-attachments/assets/5da3b797-9fc3-485b-9da3-bda363d45ba2

after

https://github.com/user-attachments/assets/e1678aa2-7367-4bb7-8b66-e1e82cd999b7

## Summary by Sourcery

Allow CFP submission and speaker profile forms to be saved as drafts without enforcing required-field validation, while preserving full validation for final submissions.

New Features:
- Introduce a non-strict mode for CFP, submission, and profile forms to relax required fields and validators when saving drafts.
- Support saving partially completed CFP submission steps as drafts, including for unauthenticated users, with a default draft proposal title when missing.

Bug Fixes:
- Ensure wizard step completion checks and finalization logic handle draft submissions without blocking on missing required fields.

Enhancements:
- Propagate a not-strict flag through question injection and dynamic field generation to conditionally disable requirements and length/date validators.
- Show a user-facing warning when a draft is saved despite validation errors instead of treating the submission as fully invalid.
- Mount the tests directory into the Docker compose web service container to simplify running tests in development.

### CODE REVIEW

**Comment by mariobehling** [comment]
Small process note.

We have automatic Copilot PR reviews enabled on this repository. These reviews are only triggered if the contributor has GitHub Copilot enabled and an active license on their own account.

Please enable Copilot in your GitHub settings if you have access. In many regions, free licenses are available through educational institutions or developer programs. Enabling Copilot helps us speed up the auto review process and reduces manual review overhead for the core team.

The team will review your PR. Thank you for your contribution and cooperation.

**Comment by shivam-pawar-7217** [comment]
Nice work on the draft feature!

Quick question - the PR says it supports unauthenticated users, but looking at UserStep.done() in flow.py, it seems like UserForm still needs login data even when draft=True. 

How would anonymous users save drafts?

**Comment by sarafarajnasardi** [comment]
@shivam-pawar-7217 The mention of 'unauthenticated users' in the automated PR summary by Sourcery was an error account creation is still intended to be mandatory for saving drafts in our flow.

**Review by ArnavBallinCode — CHANGES_REQUESTED** [review]
please consider copilot review

**Review by MukundC25 — COMMENTED** [review]
@sarafarajnasardi Address the AI reviews

**Comment by sarafarajnasardi** [comment]
@MukundC25 ok

**Review by mariobehling — CHANGES_REQUESTED** [review]
Please resolve conflicts and finalize this PR first before opening more PRs. For new contributors, we prefer to have no more than three open PRs per person at a time. Ideally, focus on one PR, get it reviewed and merged, then move on to the next. This keeps feedback cycles manageable and helps you iterate more effectively.

**Review by ArnavBallinCode — CHANGES_REQUESTED** [review]
Please address AI reviews and confirm functionality after changes.

**Comment by sarafarajnasardi** [comment]
@ArnavBallinCode I have confirmed the functionality it's working properly as expected...

File: app/eventyay/cfp/forms/cfp.py [hunk 1]
```diff
@@ -12,7 +12,11 @@ class CfPFormMixin:
     """
 
     def __init__(self, *args, field_configuration=None, **kwargs):
+        self.not_strict = kwargs.pop('not_strict', False)
         super().__init__(*args, **kwargs)
+        if self.not_strict:
+            for field in self.fields.values():
+                field.required = False
```
[the remark this answers is not in the document]
Comment (ArnavBallinCode) [reply app/eventyay/cfp/forms/cfp.py]: your `not_strict` override is running before `field_configuration` and the `field_configuration` data is processed after the `not_strict` loop . Please address this review

File: app/eventyay/cfp/flow.py [hunk 2]
```diff
@@ -304,6 +318,18 @@ def post(self, request):
...
             if error_message:
+                if action == 'draft':
+                    messages.error(
+                        self.request,
+                        _(
+                            'Your draft could not be saved because there are errors in your submission. '
+                            'Your current inputs on this step have been kept; please review and fix the errors below.'
+                        ),
+                    )
+                    self.set_data(form.cleaned_data)  # Keep valid input in the session even if the form is invalid
```
Comment (ArnavBallinCode) [inline app/eventyay/cfp/flow.py]: `form.cleaned_data` only contains fields that passed their own field-level validation. Fields with any error will be excluded ?

**Review by ArnavBallinCode — CHANGES_REQUESTED** [review]
I am unable to update my draft and then save it again , please fix this and address the remaining issues. 

Image

File: app/eventyay/cfp/forms/cfp.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ArnavBallinCode) [reply app/eventyay/cfp/forms/cfp.py]: please un resolve and address this .

**Comment by sarafarajnasardi** [comment]
@ArnavBallinCode here abstract is required field you should fill that to submit the proposal...

File: app/eventyay/cfp/flow.py (hunk 2 above)
Comment (sarafarajnasardi) [reply app/eventyay/cfp/flow.py]: Yes, ``form.cleaned_data`` only contains successfully validated fields, so any fields with errors are excluded.

File: app/eventyay/cfp/forms/cfp.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sarafarajnasardi) [reply app/eventyay/cfp/forms/cfp.py]: ok

**Comment by sarafarajnasardi** [comment]
@ArnavBallinCode fixed

**Review by ArnavBallinCode — CHANGES_REQUESTED** [review]
Please address the copilot suggestions

**Review by ArnavBallinCode — CHANGES_REQUESTED** [review]
In order to save the draft for the entire submission, the user should be able to continue to the next page and fill in whatever needed on the 2nd page too, but this pr doesn't allow users to press the "continue" button / go to the next page and fill ind details to save as a draft without having filled all the required fields.
Please fix this !

**Comment by sarafarajnasardi** [comment]
@ArnavBallinCode Thanks for addressing the bug....I have fixed that....

**Review by ArnavBallinCode — CHANGES_REQUESTED** [review]
There is still a confusing issue in the draft submission flow when any required field is left empty.

#### Steps to reproduce
1. Fill the CFP form partially, leaving at least one required field empty.
2. Continue to the next step.
3. Click **Submit proposal**.

#### Current behaviour
The user is redirected back to Page 1 without any warning or validation message.

#### Issue
This creates a silent loop where users cannot tell:
- which field is missing,
- whether the draft was saved,
- or why they were sent back.

**Comment by sarafarajnasardi** [comment]
I have fixed that in my latest commit....Thanks for pointing this out...

**Review by ArnavBallinCode — CHANGES_REQUESTED** [review]
You are completely changing the behaviour, you made those fields optional but they are required, please test your pr thoroughly using different scenarios, keep draft till you are not sure.

Image

Image

**Comment by sarafarajnasardi** [comment]
Thank you for the feedback! Just to clarify .... the required field validation still works correctly on draft form submission, the fields still throw errors if left empty. The change is only in the UI label display. But you're right, I should test all scenarios thoroughly before marking it ready. I'll test edge cases and update you shortly

**Comment by ArnavBallinCode** [comment]
> is only in the UI label display.

this should not be changed, it may cause confusion to the people if the field is actually optional or required.

**Comment by sarafarajnasardi** [comment]
You are Right....I've already reverted the change, just testing a few scenarios to make sure everything works correctly. Will update you shortly...

**Comment by sarafarajnasardi** [comment]
Hi @ArnavBallinCode,
I’ve applied all the suggested changes and re-tested the draft and submission flows.
I tried different scenarios to make sure things are working fine:
- Draft save works even if fields are empty or partially filled
- Browser validation is not blocking in draft, but works properly during submit
- Custom validations are not breaking draft flow
- On final submit, missing fields are showing proper errors
- Also checked the earlier issue where it was redirecting to Step 1 without any message....now it redirects correctly and shows validation errors on the exact step

Also tested back navigation and basic auth cases, everything looks stable.
Please let me know if I missed anything or if you want me to test any other cases.

File: app/eventyay/cfp/views/wizard.py [hunk 3]
```diff
@@ -68,13 +64,23 @@ def dispatch(self, request, *args, **kwargs):
...
             return result
+        # When "Continue" is clicked on the last step, FormFlowStep.post returns None
+        # (no next URL). We need to call done() to finalise. In draft mode, done(draft=True)
+        # is called; otherwise done() enforces full validation.
+        action = request.POST.get('action', 'submit')
+        is_draft_mode = request.GET.get('draft') == '1'
+        if request.method == 'POST' and action == 'submit' and result is None:
+            # On the final step (no next step), "submit" means final submission.
+            # This must be STRICT (draft=False) to ensure no required fields are missing,
+            # even if the user were previously navigating in draft mode.
```
Comment (ArnavBallinCode) [inline app/eventyay/cfp/views/wizard.py]: same here

File: app/eventyay/cfp/flow.py [hunk 4]
```diff
@@ -299,7 +313,46 @@ def post(self, request):
             prev_url = self.get_prev_url(request)
             return redirect(prev_url) if prev_url else redirect(request.path)
 
-        # For "submit" and "draft" actions, validate as before
+        # For "submit" action (Continue button), we always allow advancing to the next step
+        # to ensure a smooth navigation UX. If the form is invalid (e.g. missing required
+        # fields), we save the partial data and transition the flow to draft mode by
+        # appending ?draft=1 to the next URL. This allows users to skip fields and
+        # return later, while ensuring that the final "Submit" check will still block.
```
Comment (ArnavBallinCode) [inline app/eventyay/cfp/flow.py]: This comment is just unnecessary, follow repository coding instructions.

**Review by ArnavBallinCode — CHANGES_REQUESTED** [review]

File: app/eventyay/cfp/forms/cfp.py [hunk 5]
```diff
@@ -22,6 +23,34 @@ def __init__(self, *args, field_configuration=None, **kwargs):
...
+                    # but keep field.required = True so the * indicator stays visible
+                    field.draft_optional = True
+
+    def _clean_fields(self):
+        """Temporarily relax required fields for draft saves.
+
+        During validation we flip draft_optional fields to required=False so
+        Django's blank-value check doesn't produce an error.  We restore the
+        original value immediately after so that *rendering* still shows the
+        asterisk (field.required is used by widgets / label_tag).
```
Comment (ArnavBallinCode) [inline app/eventyay/cfp/forms/cfp.py]: this also can be reduced.

**Review by ArnavBallinCode — CHANGES_REQUESTED** [review]

**Review by ArnavBallinCode — COMMENTED** [review]
Please address copilot reviews as well

File: app/eventyay/cfp/forms/cfp.py (hunk 5 above)
Comment (sarafarajnasardi) [reply app/eventyay/cfp/forms/cfp.py]: ok

File: app/eventyay/person/forms/profile.py [hunk 6]
```diff
@@ -133,6 +136,7 @@ def __init__(self, *args, name=None, enforce_account_name_match=False, **kwargs)
             event=self.event,
             speaker=self.user,
             readonly=read_only,
+            not_strict=self.not_strict,
```
Comment (ArnavBallinCode) [inline app/eventyay/person/forms/profile.py:154]: Previously, if talks_published was False, both agenda AND cfp pages returned 404. Now CFP pages are always accessible regardless of this setting.

**Review by ArnavBallinCode — COMMENTED** [review]
Please confirm the following behaviour

File: app/eventyay/person/forms/profile.py (hunk 6 above)
Comment (sarafarajnasardi) [reply app/eventyay/person/forms/profile.py:154]: Image


CFP pages are not accessible if talk_published is false

**Review by Sak1012 — APPROVED** [review]

**Review by ArnavBallinCode — APPROVED** [review]

File: app/eventyay/common/forms/mixins.py [hunk 7]
```diff
@@ -151,10 +155,29 @@ def validate_field_length(value, min_length, max_length, count_in):
...
-        qs = TalkQuestion.all_objects.filter(event=event, active=True, target=target)
+        qs = TalkQuestion.all_objects.filter(
+            event=event,
+            active=True,
+            is_imported=False,
+            target=target,
+        )
         return qs.order_by('position')
 
+<<<<<<< cfp-draft-submission
```
Comment (ArnavBallinCode) [inline app/eventyay/common/forms/mixins.py]: Looks like unresolved merge conflict markers were accidentally committed here:

```python
<<<<<<< cfp-draft-submission
=======
>>>>>>> dev

**Review by ArnavBallinCode — CHANGES_REQUESTED** [review]

File: app/eventyay/common/forms/mixins.py (hunk 7 above)
Comment (sarafarajnasardi) [reply app/eventyay/common/forms/mixins.py]: fixed

**Review by Saksham-Sirohi — CHANGES_REQUESTED** [review]
Image

@Sak1012 @sarafarajnasardi the submission leaves two different messages, one indicating success and one error, which may cause confusion to the user. Furthermore, people can still save drafts without filling a single field. Is this desired? I get skipping mandatory fields, but a draft with no detail holds no point; they might as well try a new submission at that point

**Comment by sarafarajnasardi** [comment]
> Image
> @Sak1012 @sarafarajnasardi the submission leaves two different messages, one indicating success and one error, which may cause confusion to the user. Furthermore, people can still save drafts without filling a single field. Is this desired? I get skipping mandatory fields, but a draft with no detail holds no point; they might as well try a new submission at that point

Yeah that makes sense.... a completely empty draft isn't useful. I've fixed the dual message issue. For the empty draft behavior, could you check with @mario on what the minimum should be (e.g. at least a title)? I'll update it based on that.

File: app/eventyay/cfp/flow.py [hunk 8]
```diff
@@ -302,7 +316,40 @@ def post(self, request):
...
+                    self.set_data(form.cleaned_data)
+                if form.files:
+                    self.set_files(form.files)
+
+            next_step = self.get_next_applicable(request)
+            if next_step:
+                query = {}
+                if not is_valid:
+                    query['draft'] = 1
+                return redirect(next_step.get_step_url(request, query=query))
```
[the remark this answers is not in the document]
Comment (sarafarajnasardi) [reply app/eventyay/cfp/flow.py:344]: this is desired

**Comment by Saksham-Sirohi** [comment]
> > Image
> 
> > @Sak1012 @sarafarajnasardi the submission leaves two different messages, one indicating success and one error, which may cause confusion to the user. Furthermore, people can still save drafts without filling a single field. Is this desired? I get skipping mandatory fields, but a draft with no detail holds no point; they might as well try a new submission at that point
> 
> 
> 
> Yeah that makes sense.... a completely empty draft isn't useful. I've fixed the dual message issue. For the empty draft behavior, could you check with @mario on what the minimum should be (e.g. at least a title)? I'll update it based on that.
> 
> 
> 
> 

Instead of saving blanks there should be at least one entry done by user in order to trigger this @Sak1012 @mariobehling is this the correct direction

**Review by ArnavBallinCode — CHANGES_REQUESTED** [review]
please resolve merge conflicts

**Comment by sarafarajnasardi** [comment]
https://github.com/user-attachments/assets/d52d5e61-b7d4-4d2b-b4ac-a0e954fb164c

@Sak1012 @Saksham-Sirohi I have updated my pr as per requested changes.....

**Review by Saksham-Sirohi — CHANGES_REQUESTED** [review]
Add a draft proposal date so that people can track better

File: app/eventyay/cfp/flow.py [hunk 9]
```diff
@@ -413,9 +462,81 @@ def get_form_initial(self):
...
+    def get_form(self, from_storage=False, not_strict=None):
+        form = super().get_form(from_storage=from_storage, not_strict=not_strict)
+        if from_storage and not_strict:
+            changed = False
+
+            if not form.data.get('title'):
+                # Inject a placeholder title so the DB non-blank constraint is satisfied.
+                # This value is only stored internally; when the user edits the draft,
+                # the title field will appear empty (see below).
+                form.data['title'] = str(_(self.DRAFT_TITLE_PLACEHOLDER))
```
Comment (ArnavBallinCode) [inline app/eventyay/cfp/flow.py]: This may break if the draft is created under one language and later edited under another. 

The placeholder would no longer match and could remain as a real title.

Would it be safer to use a language-independent internal sentinel value such as:

```python
DRAFT_TITLE_PLACEHOLDER = '__AUTO_DRAFT_TITLE__'
```
and only translate at the display/UI layer?

**Review by ArnavBallinCode — CHANGES_REQUESTED** [review]

File: app/eventyay/cfp/flow.py (hunk 9 above)
Comment (sarafarajnasardi) [reply app/eventyay/cfp/flow.py]: I have updated the PR as per the suggestion....Thanks

**Comment by sarafarajnasardi** [comment]
image

@Saksham-Sirohi  I have added the date when the proposal was last modified.

**Review by Saksham-Sirohi — APPROVED** [review]
Think this is good enough for this pull request. If anything else is required, it can be a follow-up LGTM

**Review by Rachit7168 — APPROVED** [review]
LGTM , Tested 

https://github.com/user-attachments/assets/e02acfd5-2242-4e3f-9d09-0d2a2430754b

**Review by Rachit7168 — CHANGES_REQUESTED** [review]
https://github.com/user-attachments/assets/cc198036-8f57-49c6-8313-24ef9ae58012

Please check there is an issue with radio one button I removed the selected option but it still selects and shows one and save as draft !

Also draft save are shown as __AUTO_DRAFT_TITLE__ when opened the draft maybe change it to something more enterprise type like draft proposal 1 , draft proposal 2

**Review by Rachit7168 — CHANGES_REQUESTED** [review]
https://github.com/user-attachments/assets/8ef27705-a461-47d2-a576-2d5c4bb3ad0d

Please check I'm getting two message , one would be better and less confusing 

https://github.com/user-attachments/assets/3064cc21-3ee6-4534-aa4f-931fffa72e8b

After a draft is saved and i open and make change and try to save the draft it throws error and also the draft proposal title become blank from **“__AUTO_DRAFT_TITLE__”** to **“”**

**Comment by sarafarajnasardi** [comment]
Closing in favor of #3652
